### PR TITLE
Add api for certificate revocation

### PR DESCRIPTION
**Pull request by remotelyliving**
First off, thanks for any time you can spend reviewing this PR! We greatly appreciate the work done on this lib and would love to give back. Revoke kind of helps round out the flow for managing a certificate.

I took some ideas from the python library as there is not much documentation on revoke, but it's simple enough. You can revoke once, and then subsequent attempts error our with a 409. 

You must use the same account keypair used to request the certificate to sign the request.

That's about it!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,4 @@ tests/Cli/Fixtures/challenges/.well-known/acme-challenge/*
 !tests/Cli/Fixtures/challenges/.well-known/acme-challenge/.gitkeep
 tests/Core/Fixtures/challenges/.well-known/acme-challenge/*
 !tests/Core/Fixtures/challenges/.well-known/acme-challenge/.gitkeep
+.idea
```
Comment (tgalopin) [inline .gitignore]: This should be in your global gitignore (have a look at https://gist.github.com/subfuzion/db7f57fff2fb6998a16c).

File: src/Cli/Command/RevokeCommand.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    protected function configure()
+    {
+        $this->setName('revoke')
+            ->setDefinition([
+                new InputArgument('domain', InputArgument::REQUIRED, 'The domain revoke a certificate for'),
+                new InputArgument('reason-code', InputOption::VALUE_OPTIONAL, 'The reason code for revocation'),
+            ])
+            ->setDescription('Revoke a SSL certificate for a domain')
+            ->setHelp(<<<'EOF'
+The <info>%command.name%</info> command revoke a previously obtained certificate for a given domain
```
Comment (tgalopin) [inline src/Cli/Command/RevokeCommand.php]: If there is some kind of documentation for possible reason codes, that would be great to put it in the command description so users are able to find them easily :) .

File: src/Cli/Command/RevokeCommand.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class RevokeCommand extends AbstractCommand
+{
+    /**
+     * @var RepositoryInterface
+     */
+    private $repository;
```
Comment (tgalopin) [inline src/Cli/Command/RevokeCommand.php]: You don't really need properties, do you? I would be in favor of using local variables in the execute method.

File: src/Cli/Command/RevokeCommand.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $this->repository = $this->getRepository();
+        $this->client = $this->getClient();
+        $this->actionHandler = $this->getActionHandler();
+
+        $domain = (string) $input->getArgument('domain');
+        $reasonCode = (int) $input->getArgument('reason-code');
+
+        if (!$this->repository->hasDomainCertificate($domain)) {
+
+            $this->error("Certificate for {$domain} not found locally");
```
Comment (tgalopin) [inline src/Cli/Command/RevokeCommand.php]: You should use a normal concatenation (`'Certificate for '.$domain.' not found locally'`)

**Comment by remotelyliving** [comment]
Thank you so much @tgalopin, great suggestions all around. I provided some more context around reason types. I'm having a tough time getting the test suite to run...it never exposes port 14000. So revoke in the `AcmeClientTest::testFullProcess()` is failing. I don't know what the cert looks like that it gets back, so that could be the issue if it's more of a stub and not a real cert. 

Revoke works IRL with let's encrypt so I can't tell if the test failing is just a difference of environment or something significant. 

```
Christians-MacBook-Pro:acmephp chthomas$ docker ps
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS                  NAMES
cc4c0a16a01e        acmephp/testing-ca:2.0   "/go/bin/pebble"         18 minutes ago      Up 18 minutes                              acme_boulder
de47444fd767        atmoz/sftp               "/entrypoint acmephp…"   18 minutes ago      Up 18 minutes       0.0.0.0:8022->22/tcp   acme_sftpChristians-MacBook-Pro:acmephp chthomas$ docker ps
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS                  NAMES
cc4c0a16a01e        acmephp/testing-ca:2.0   "/go/bin/pebble"         18 minutes ago      Up 18 minutes                              acme_boulder
de47444fd767        atmoz/sftp               "/entrypoint acmephp…"   18 minutes ago      Up 18 minutes       0.0.0.0:8022->22/tcp   acme_sftp```

Anyways, I'll pick back up with you tomorrow. Thanks again for helping out with this! I'm excited to get to work together.

File: src/Core/AcmeClient.php [hunk 5]
```diff
@@ -214,6 +216,38 @@ public function requestCertificate($domain, CertificateRequest $csr, $timeout =
...
+    /**
+     * {@inheritdoc}
+     */
+    public function revokeCertificate(Certificate $certificate, RevocationReason $revocationReason = null)
+    {
+        if (null === $revocationReason) {
+            $revocationReason = RevocationReason::createDefaultReason();
+        }
+
+        $formatted = str_ireplace('-----BEGIN CERTIFICATE-----', '', $certificate);
```
[the remark this answers is not in the document]
Comment (remotelyliving) [reply src/Core/AcmeClient.php]: `Certificate::getPem()` conveys the intent *and* the format.

File: src/Core/Protocol/RevocationReason.php [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            $formatted[] = $reason.' - '.$label;
+        }
+
+        return $formatted;
+    }
+
+    /**
+     * @return array
+     */
+    public static function getReasonLabelMap()
```
[the remark this answers is not in the document]
Comment (jderusse) [reply src/Core/Protocol/RevocationReason.php]: Good catch !

**Comment by remotelyliving** [comment]
@tgalopin I am so sorry to be a bother, but I can't seem to reconcile the differences in the test environment with production. Revoke works with LE production environments. I couldn't get the test setup to work locally but I did manage to get the latest version of Pebble running. My changes + the most recent version of Pebble pass. Do you think we should just remove the test for now? I know that's not optimal, but I'm wondering if the differences in environments is more noise than it's worth.

**Comment by jderusse** [comment]
@remotelyliving do you think it's possible to update acmephp/testing-ca to the last version of pebble and fix your test suit?

**Comment by tgalopin** [comment]
I'd be glad to give hand here to finish the PR. @remotelyliving are you still interested :) ? If so, ping me here or on the Symfony Slack (https://symfony.com/support)!

**Comment by tgalopin** [comment]
Hey @remotelyliving !

I fixed the final small issues in https://github.com/acmephp/acmephp/pull/124 so I'm going to close this PR (but you name will still be present in the commits of the other PR ;) ).

Thanks a lot for your work!

**Comment by tgalopin** [comment]
Many thanks to you @remotelyliving, revocation is now in AcmePHP 1.1 :) ! https://twitter.com/acme_php/status/1061228086073704448

**Comment by remotelyliving** [comment]
@tgalopin amazing!! My apologies for not responding sooner. Our company switched to private git and I turned my notifications off for public git...but forgot people might actually be trying to get in touch with me! They're back on now. I'd love to contribute further and again, many thanks for maintaining this library to begin with!

**Pull request closed without merging**